### PR TITLE
Evaluate operations against a policy enforcer over NATS

### PR DESCRIPTION
**Pull request by brooksmtownsend**
Implementation of the host-side of the RFC #439 

This PR includes a new module, `HostCore.Policy.Manager` which is responsible for taking a source, target, and action and querying a policy topic for an assertive allow or deny. This only occurs when a host is configured to use a policy topic, `WASMCLOUD_POLICY_TOPIC`, and quickly returns an allow in the case that no policy topic is configured.

In the case that a policy query times out (e.g. isn't answered by a policy enforcer) the wasmCloud host fails **closed** and will **not allow** the request to go through. Note that this means that a host configured with a policy topic that is invalid, or when a policy server goes down, no actors or providers can be started that haven't been previously authorized and no invocations can go through that haven't already happened between existing resources (cc @stevelr @autodidaddict)

When policy requests are denied due to invalid structures or timing out, those results will not be cached

### Some side effects that might not be initially obvious and need some extra thought
1. If a policy topic is configured on accident or with a typo, the first request to do any new action (start a new actor, perform an invocation, etc) will take 2 seconds longer as it waits to time out. This timeout probably needs to be adjusted as 2 seconds is enough to fail any invocation with the default RPC timeout (assuming it's coming from a provider)
2. If a policy request is invalid (e.g. missing an issuer key), it's allowed without consulting the policy engine. We control writing these requests in the host so it _shouldn't_ be an issue, but if an actor is missing an issuer I could see the argument to deny it.

### Still TODO:
- [x] Implement `WASMCLOUD_POLICY_CHANGES_TOPIC` to invalidate portions of the cache
- [x] Implement policy evaluation for `start_provider` and `perform_invocation`

### CODE REVIEW

File: host_core/lib/host_core/actors/actor_module.ex [hunk 1]
```diff
@@ -405,7 +406,66 @@ defmodule HostCore.Actors.ActorModule do
...
+           issued_on: source_claims[:iat],
+           expires_in_mins: source_claims[:exp]
+         },
+         %{
+           public_key: target["public_key"],
+           contract_id: target["contract_id"],
+           link_name: target["link_name"],
+           issuer: target_claims[:iss]
+         },
+         "perform_invocation"
```
Comment (autodidaddict) [inline host_core/lib/host_core/actors/actor_module.ex:434]: we should probably put these action type constants into module consts

File: host_core/lib/host_core/actors/actor_module.ex [hunk 2]
```diff
@@ -405,7 +406,66 @@ defmodule HostCore.Actors.ActorModule do
...
+    {:error, "actor is missing capability claims"}
+  end
+
+  # Deny invocation if policy enforcer does not permit it
+  defp perform_invocation(
+         {{_agent, true}, %{permitted: false} = policy_res},
+         _operation,
+         _payload
+       ) do
+    message = Map.get(policy_res, :message, "reason not specified")
```
Comment (autodidaddict) [inline host_core/lib/host_core/actors/actor_module.ex:458]: so many opportunities to provide a hilarious default message here.

Comment (autodidaddict) [reply host_core/lib/host_core/actors/actor_module.ex:458]: `Ye have been weighed, you have been measured, and ye have been found wanting`

File: host_core/lib/host_core/policy/manager.ex [hunk 3]
```diff
@@ -0,0 +1,160 @@
+defmodule HostCore.Policy.Manager do
+  @moduledoc false
+  require Logger
+  # TODO: Gnat.server? Listen for
```
Comment (autodidaddict) [inline host_core/lib/host_core/policy/manager.ex:4]: does this comment still apply?

**Review by autodidaddict — COMMENTED** [review]
looking good so far

File: host_core/lib/host_core/policy/manager.ex [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+         expires_in_mins: _expires
+       }) do
+    :ok
+  end
+
+  defp validate_source(source) when is_map(source) do
+    # Narrow down missing fields by removing present fields from the list
+    missing_fields =
+      [:public_key, :capabilities, :issuer, :issued_on, :expires_in_mins]
+      |> Enum.reject(fn required_field -> Map.get(source, required_field) != nil end)
```
Comment (connorsmith256) [inline host_core/lib/host_core/policy/manager.ex:107]: `reject` with `!= nil` hurts my brain. Why not `filter` with `== nil`?

File: host_core/lib/host_core/policy/manager.ex (hunk 3 above)
Comment (brooksmtownsend) [reply host_core/lib/host_core/policy/manager.ex]: nope! removed

File: host_core/lib/host_core/policy/manager.ex (hunk 4 above)
Comment (brooksmtownsend) [reply host_core/lib/host_core/policy/manager.ex]: good call!

File: host_core/test/host_core/policy_test.exs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+      )
+
+    # Ensure the host doesn't start the actor that's denied
+    assert !(HostCore.Actors.ActorSupervisor.all_actors()
+             |> Map.keys()
+             |> Enum.any?(fn public_key ->
+               public_key == "MB2ZQB6ROOMAYBO4ZCTFYWN7YIVBWA3MTKZYAQKJMTIHE2ELLRW2E3ZW"
+             end))
+
+    # # Ensure the host doesn't start the provider that's denied
```
Comment (connorsmith256) [inline host_core/test/host_core/policy_test.exs:207]: Supernit: double comment

**Review by connorsmith256 — APPROVED** [review]

**Comment by brooksmtownsend** [comment]
> @brooksmtownsend do you know why the build is failing checks?

Yep looks like I missed a refactor point. pushed a fix 🤞🏻

**Review by connorsmith256 — APPROVED** [review]

**Review by autodidaddict — APPROVED** [review]

**Pull request merged**